File: ext/types/task_id.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

TaskIDPtr toTaskID(char* bs, int len)
{
  TaskIDPtr val = new TaskID();
  val->set_value(bs, len);
  return val;
}

int fromTaskID(TaskIDPtr p, char** poke)
{
  *poke = (char*) p->mutable_value()->data();
  return p->mutable_value()->size();
}

void destroyTaskID(TaskIDPtr p)
{
  delete p;
}

File: src/System/Mesos/Raw/TaskId.hs
module System.Mesos.Raw.TaskId where
import           System.Mesos.Internal

type TaskIDPtr = Ptr TaskID

foreign import ccall "ext/types.h toTaskID" c_toTaskID :: ToID TaskIDPtr

foreign import ccall "ext/types.h fromTaskID" c_fromTaskID :: FromID TaskIDPtr

foreign import ccall "ext/types.h destroyTaskID" c_destroyTaskID :: TaskIDPtr -> IO ()

instance CPPValue TaskID where
  marshal x = do
    (strp, l) <- cstring $ fromTaskID x
    liftIO $ c_toTaskID strp $ fromIntegral l

  unmarshal p = fmap TaskID $ do
    pp <- alloc
    len <- liftIO $ c_fromTaskID p pp
    peekCString' (pp, len)

  destroy = c_destroyTaskID
